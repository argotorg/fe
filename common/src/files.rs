use crate::utils::keccak;
use crate::Span;
use codespan_reporting as cs;
use cs::files::Error as CsError;
use std::{
    collections::HashMap,
    convert::TryInto,
    fs,
    io,
    ops::Range,
    path::Path,
};

pub struct SourceFile {
    id: SourceFileId,
    name: String,
    content: String,
    line_starts: Vec<usize>,
}

#[derive(PartialEq, Copy, Clone, Eq, Hash)]
pub struct SourceFileId(u128);

impl SourceFile {
    pub fn new(name: String, content: String) -> Self {
        let hash = keccak::full_as_bytes(content.as_bytes());
        let line_starts = cs::files::line_starts(&content).collect();
        Self {
            id: SourceFileId(u128::from_be_bytes(hash[..16].try_into().unwrap())),
            name,
            content,
            line_starts,
        }
    }

    pub fn line_index(&self, byte_index: usize) -> usize {
        self.line_starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1)
    }

    pub fn line_span(&self, line_index: usize) -> Option<Span> {
        Some(Span::new(
            *self.line_starts.get(line_index)?,
            *self.line_starts.get(line_index + 1)?,
        ))
    }
}

pub trait FileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String>;
}

pub struct OsFileLoader;

impl FileLoader for OsFileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

pub struct FileStore {
    files: HashMap<SourceFileId, SourceFile>,
    loader: Box<dyn FileLoader>,
}

impl FileStore {
    pub fn new() -> Self {
        Self {
            files: HashMap::new(),
            loader: Box::new(OsFileLoader),
        }
    }

    pub fn with_loader(loader: Box<dyn FileLoader>) -> Self {
        Self {
            files: HashMap::new(),
            loader,
        }
    }

    pub fn add_file(&mut self, path: String, content: String) -> SourceFileId {
        let file = SourceFile::new(path, content);
        let id = file.id;
        self.files.insert(id, file);
        id
    }

    pub fn load_file(&mut self, path: String) -> io::Result<(SourceFileId, String)> {
        let content = self.loader.load_file(&Path::new(&path))?;
        let id = self.add_file(path, content.clone());
        Ok((id, content))
    }

    pub fn get_file(&self, id: SourceFileId) -> Option<&SourceFile> {
        self.files.get(&id)
    }
}

impl<'a> cs::files::Files<'a> for FileStore {
    type FileId = SourceFileId;
    type Name = &'a str;
    type Source = &'a str;

    fn name(&'a self, id: SourceFileId) -> Result<Self::Name, CsError> {
        self.get_file(id)
            .map(|file| file.name.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn source(&'a self, id: SourceFileId) -> Result<Self::Source, CsError> {
        self.get_file(id)
            .map(|file| file.content.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn line_index(&'a self, id: SourceFileId, byte_index: usize) -> Result<usize, CsError> {
        Ok(self
            .get_file(id)
            .ok_or(CsError::FileMissing)?
            .line_index(byte_index))
    }

    fn line_range(&'a self, id: SourceFileId, line_index: usize) -> Result<Range<usize>, CsError> {
        let file = self.get_file(id).ok_or(CsError::FileMissing)?;
        Ok(file
            .line_span(line_index)
            .ok_or(CsError::LineTooLarge {
                given: line_index,
                max: file.line_starts.len() - 1,
            })?
            .into())
    }
}
