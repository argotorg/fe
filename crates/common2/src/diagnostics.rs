use parser::ast::SyntaxNodePtr;

use crate::InputFile;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct CompleteDiagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Span,
    pub sub_diagnostics: Vec<SubDiagnostic>,
    pub error_code: GlobalErrorCode,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct GlobalErrorCode {
    pub pass: AnalysisPass,
    pub local_code: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SubDiagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Span {
    pub file: InputFile,
    pub node: SyntaxNodePtr,
    pub origin: SpanOrigin,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum Severity {
    Error,
    Warning,
    Note,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum SpanOrigin {
    Raw,
    Expanded,
    Desugared,
}

#[repr(u16)]
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum AnalysisPass {
    Parse = 1,
    NameResolution,
    TyCheck,

    ExternalAnalysis(ExternalAnalysisKey) = u16::MAX,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ExternalAnalysisKey {
    name: String,
}
