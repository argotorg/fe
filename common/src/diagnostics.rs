use crate::files::{FileStore, SourceFileId};
pub use codespan_reporting::diagnostic::{Diagnostic as CsDiagnostic, Label, LabelStyle, Severity};
use codespan_reporting::term;
use term::termcolor::{BufferWriter, ColorChoice};

pub type Diagnostic = CsDiagnostic<SourceFileId>;

/// Print the given diagnostics to stderr.
pub fn print_diagnostics(diagnostics: &[Diagnostic], files: &FileStore) {
    let writer = BufferWriter::stderr(ColorChoice::Auto);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).unwrap();
    }
    // If we use `writer` here, the output won't be captured by rust's test system.
    eprintln!("{}", std::str::from_utf8(buffer.as_slice()).unwrap());
}

/// Format the given diagnostics as a string.
pub fn diagnostics_string(diagnostics: &[Diagnostic], files: &FileStore) -> String {
    let writer = BufferWriter::stderr(ColorChoice::Never);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).expect("failed to emit diagnostic");
    }
    std::str::from_utf8(buffer.as_slice()).unwrap().to_string()
}
