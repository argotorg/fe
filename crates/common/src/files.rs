use crate::utils::keccak;
use crate::Span;
use codespan_reporting as cs;
use cs::files::Error as CsError;
use include_dir::Dir;
use indexmap::indexmap;
use indexmap::IndexMap;
use smol_str::SmolStr;
use std::collections::BTreeMap;
use std::ops::Range;
use std::path::Path;
use std::{fs, io};

#[derive(PartialEq, Clone, Eq, Hash, Debug)]
pub struct SourceFile {
    pub id: SourceFileId,
    pub name: String,
    pub content: String,
    line_starts: Vec<usize>,
}

#[derive(PartialEq, Copy, Clone, Eq, Hash, Debug, PartialOrd, Ord, Default)]
pub struct SourceFileId(pub u128);

impl SourceFile {
    pub fn new(name: &str, content: &str) -> Self {
        // Canonicalize new line character for Windows.
        let content_canonicalized = content.replace("\r\n", "\n");
        let hash = keccak::full_as_bytes(content_canonicalized.as_bytes());

        let line_starts = cs::files::line_starts(content).collect();
        Self {
            id: SourceFileId(u128::from_be_bytes(hash[..16].try_into().unwrap())),
            name: name.to_string(),
            content: content.to_string(),
            line_starts,
        }
    }

    pub fn line_index(&self, byte_index: usize) -> usize {
        self.line_starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1)
    }

    pub fn line_span(&self, line_index: usize) -> Option<Span> {
        let end = if line_index == self.line_starts.len() - 1 {
            self.content.len()
        } else {
            *self.line_starts.get(line_index + 1)?
        };
        Some(Span::new(self.id, *self.line_starts.get(line_index)?, end))
    }
}

pub trait FileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String>;
}

pub struct OsFileLoader;

impl FileLoader for OsFileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

pub struct FileStore {
    pub files: BTreeMap<SourceFileId, SourceFile>,
    loader: Box<dyn FileLoader>,
}

impl FileStore {
    pub fn new() -> Self {
        Self {
            files: BTreeMap::new(),
            loader: Box::new(OsFileLoader),
        }
    }

    pub fn with_loader(loader: Box<dyn FileLoader>) -> Self {
        Self {
            files: BTreeMap::new(),
            loader,
        }
    }

    pub fn add_file(&mut self, path: &str, content: &str) -> SourceFileId {
        let file = SourceFile::new(path, content);
        let id = file.id;
        self.files.insert(id, file);
        id
    }

    /// Adds an included dir to the file store.
    pub fn add_included_dir(&mut self, dir: &Dir) -> Vec<SourceFileId> {
        let mut file_ids = vec![];

        for file in dir.files() {
            file_ids.push(
                self.add_file(
                    file.path()
                        .to_str()
                        .expect("cannot convert file path to string"),
                    file.contents_utf8()
                        .expect("could not get utf8 encoded file content"),
                ),
            );
        }

        for sub_dir in dir.dirs() {
            file_ids.extend(self.add_included_dir(sub_dir))
        }

        file_ids
    }

    /// Adds the included libraries to the file store and returns a mapping of
    /// library names to file ids.
    pub fn add_included_libraries(&mut self) -> IndexMap<SmolStr, Vec<SourceFileId>> {
        indexmap! {
            "std".into() => self.add_included_dir(&fe_library::STD)
        }
    }

    pub fn load_file(&mut self, path: &str) -> io::Result<(String, SourceFileId)> {
        let content = self.loader.load_file(Path::new(&path))?;
        let id = self.add_file(path, &content);
        Ok((content, id))
    }

    pub fn get_file(&self, id: SourceFileId) -> Option<&SourceFile> {
        self.files.get(&id)
    }

    pub fn all_files(&self) -> Vec<SourceFileId> {
        self.files.keys().copied().collect()
    }
}

impl<'a> cs::files::Files<'a> for FileStore {
    type FileId = SourceFileId;
    type Name = &'a str;
    type Source = &'a str;

    fn name(&'a self, id: SourceFileId) -> Result<Self::Name, CsError> {
        self.get_file(id)
            .map(|file| file.name.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn source(&'a self, id: SourceFileId) -> Result<Self::Source, CsError> {
        self.get_file(id)
            .map(|file| file.content.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn line_index(&'a self, id: SourceFileId, byte_index: usize) -> Result<usize, CsError> {
        Ok(self
            .get_file(id)
            .ok_or(CsError::FileMissing)?
            .line_index(byte_index))
    }

    fn line_range(&'a self, id: SourceFileId, line_index: usize) -> Result<Range<usize>, CsError> {
        let file = self.get_file(id).ok_or(CsError::FileMissing)?;
        Ok(file
            .line_span(line_index)
            .ok_or(CsError::LineTooLarge {
                given: line_index,
                max: file.line_starts.len() - 1,
            })?
            .into())
    }
}

impl Default for FileStore {
    fn default() -> Self {
        Self::new()
    }
}
