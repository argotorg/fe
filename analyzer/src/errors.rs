//! Semantic errors.

use ansi_term::Color::Red;
use fe_common::diagnostics::Diagnostic;
use fe_parser::node::Span;

/// Error to be returned from APIs that should reject duplicate definitions
#[derive(Debug)]
pub struct AlreadyDefined;

#[derive(Debug)]
pub struct AnalyzerError {
    pub diagnostics: Vec<Diagnostic>,
    pub classic: Option<SemanticError>,
}

/// Errors for things that may arise in a valid Fe AST.
#[derive(Debug, PartialEq)]
pub enum ErrorKind {
    CannotMove,
    NotCallable,
    NotSubscriptable,
    SignedExponentNotAllowed,
    TypeError,
    UndefinedValue,
    Fatal,
}

#[derive(Debug, PartialEq)]
pub struct SemanticError {
    pub kind: ErrorKind,
    /// A sequence of nested spans containing the error's origin in the source
    /// code.
    pub context: Vec<Span>,
}

impl SemanticError {
    pub fn fatal() -> Self {
        SemanticError {
            kind: ErrorKind::Fatal,
            context: vec![],
        }
    }

    /// Create a new error with kind `NotSubscriptable`
    pub fn not_subscriptable() -> Self {
        SemanticError {
            kind: ErrorKind::NotSubscriptable,
            context: vec![],
        }
    }

    /// Create a new error with kind `SignedExponentNotAllowed`
    pub fn signed_exponent_not_allowed() -> Self {
        SemanticError {
            kind: ErrorKind::SignedExponentNotAllowed,
            context: vec![],
        }
    }

    /// Create a new error with kind `UndefinedValue`
    pub fn undefined_value() -> Self {
        SemanticError {
            kind: ErrorKind::UndefinedValue,
            context: vec![],
        }
    }

    /// Create a new error with kind `TypeError`
    pub fn type_error() -> Self {
        SemanticError {
            kind: ErrorKind::TypeError,
            context: vec![],
        }
    }

    /// Create a new error with kind `CannotMove`
    pub fn cannot_move() -> Self {
        SemanticError {
            kind: ErrorKind::CannotMove,
            context: vec![],
        }
    }

    /// Create a new error with kind `NotCallable`
    pub fn not_callable() -> Self {
        SemanticError {
            kind: ErrorKind::NotCallable,
            context: vec![],
        }
    }

    /// Maps the error to a new error that contains the given span in its
    /// context.
    pub fn with_context(mut self, span: Span) -> Self {
        self.context.push(span);
        self
    }

    /// Formats the error using the source code.
    ///
    /// The string will contain the error kind, line number, and surrounding
    /// code.
    pub fn format_user(&self, src: &str) -> String {
        let line = if let Some(span) = self.context.first() {
            src[..span.start].lines().count()
        } else {
            0
        };

        let context = match (self.context.get(0), self.context.get(1)) {
            (Some(inner), Some(outer)) => {
                let first_part = src[outer.start..inner.start].to_string();
                let middle_part = Red.paint(&src[inner.start..inner.end]).to_string();
                let last_part = src[inner.end..outer.end].to_string();

                format!("{}{}{}", first_part, middle_part, last_part)
            }
            (Some(span), None) => src[span.start..span.end].to_string(),
            (_, _) => "no error context available".to_string(),
        };

        format!("{:?} on line {}\n{}", self.kind, line, context)
    }
}
