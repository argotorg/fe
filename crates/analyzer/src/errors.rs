//! Semantic errors.

use fe_common::diagnostics::Diagnostic;

/// Error to be returned from APIs that should reject duplicate definitions
#[derive(Debug)]
pub struct AlreadyDefined;

/// Error to be returned when otherwise no meaningful information can be returned
#[derive(Debug)]
pub struct FatalError;

impl FatalError {
    /// Create a `FatalError` instance.
    pub fn new() -> FatalError {
        // This is primarly a hook for debugging. If we hit a FatalError for which we haven't
        // yet assigned a proper user error, we can place a panic here to see where the fatal
        // error originates.
        FatalError
    }
}

impl Default for FatalError {
    fn default() -> Self {
        Self::new()
    }
}

/// Error indicating that a value can not move between memory and storage
#[derive(Debug)]
pub struct CannotMove;

/// Error indicating that a value has the wrong type
#[derive(Debug)]
pub struct TypeError;

/// Errors that can result from indexing
#[derive(Debug, PartialEq)]
pub enum IndexingError {
    WrongIndexType,
    NotSubscriptable,
}

/// Errors that can result from a binary operation
#[derive(Debug, PartialEq)]
pub enum BinaryOperationError {
    TypesNotEqual,
    TypesNotNumeric,
    RightTooLarge,
    RightIsSigned,
    NotEqualAndUnsigned,
}

#[derive(Debug)]
pub struct AnalyzerError(pub Vec<Diagnostic>);
