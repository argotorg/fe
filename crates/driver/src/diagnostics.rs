use std::ops::Range;

use camino::Utf8Path;
use codespan_reporting as cs;
use common::{
    diagnostics::{LabelStyle, Severity},
    InputDb, InputFile,
};
use cs::{diagnostic as cs_diag, files as cs_files};
use hir_analysis::diagnostics::{DiagnosticVoucher, SpannedHirAnalysisDb};

pub trait ToCsDiag {
    fn to_cs(&self, db: &dyn SpannedInputDb) -> cs_diag::Diagnostic<InputFile>;
}

pub trait SpannedInputDb: SpannedHirAnalysisDb + InputDb {}
impl<T> SpannedInputDb for T where T: SpannedHirAnalysisDb + InputDb {}

impl<T> ToCsDiag for T
where
    T: for<'db> DiagnosticVoucher<'db>,
{
    fn to_cs(&self, db: &dyn SpannedInputDb) -> cs_diag::Diagnostic<InputFile> {
        let complete = self.to_complete(db);

        let severity = convert_severity(complete.severity);
        let code = Some(complete.error_code.to_string());
        let message = complete.message;

        let labels = complete
            .sub_diagnostics
            .into_iter()
            .filter_map(|sub_diag| {
                let span = sub_diag.span?;
                match sub_diag.style {
                    LabelStyle::Primary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Primary, span.file, span.range)
                    }
                    LabelStyle::Secondary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Secondary, span.file, span.range)
                    }
                }
                .with_message(sub_diag.message)
                .into()
            })
            .collect();

        cs_diag::Diagnostic {
            severity,
            code,
            message,
            labels,
            notes: complete.notes,
        }
    }
}

fn convert_severity(severity: Severity) -> cs_diag::Severity {
    match severity {
        Severity::Error => cs_diag::Severity::Error,
        Severity::Warning => cs_diag::Severity::Warning,
        Severity::Note => cs_diag::Severity::Note,
    }
}

#[salsa::tracked(return_ref)]
pub fn file_line_starts(db: &dyn SpannedHirAnalysisDb, file: InputFile) -> Vec<usize> {
    cs::files::line_starts(file.contents(db).text(db)).collect()
}

pub struct CsDbWrapper<'a>(pub &'a dyn SpannedHirAnalysisDb);

impl<'db> cs_files::Files<'db> for CsDbWrapper<'db> {
    type FileId = InputFile;
    type Name = &'db Utf8Path;
    type Source = &'db str;

    fn name(&'db self, file_id: Self::FileId) -> Result<Self::Name, cs_files::Error> {
        Ok(file_id.path(self.0))
    }

    fn source(&'db self, file_id: Self::FileId) -> Result<Self::Source, cs_files::Error> {
        Ok(file_id.contents(self.0).text(self.0))
    }

    fn line_index(
        &'db self,
        file_id: Self::FileId,
        byte_index: usize,
    ) -> Result<usize, cs_files::Error> {
        let starts = file_line_starts(self.0, file_id);
        Ok(starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1))
    }

    fn line_range(
        &'db self,
        file_id: Self::FileId,
        line_index: usize,
    ) -> Result<Range<usize>, cs_files::Error> {
        let line_starts = file_line_starts(self.0, file_id);

        let start = *line_starts
            .get(line_index)
            .ok_or(cs_files::Error::LineTooLarge {
                given: line_index,
                max: line_starts.len() - 1,
            })?;

        let end = if line_index == line_starts.len() - 1 {
            file_id.contents(self.0).text(self.0).len()
        } else {
            *line_starts
                .get(line_index + 1)
                .ok_or(cs_files::Error::LineTooLarge {
                    given: line_index,
                    max: line_starts.len() - 1,
                })?
        };

        Ok(Range { start, end })
    }
}
