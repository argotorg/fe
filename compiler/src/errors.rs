//! Errors returned by the compilers and ABI builder.

use fe_analyzer::errors::SemanticError;
use fe_common::diagnostics::Diagnostic;
use once_cell::sync::Lazy;
use std::borrow::Cow;
use std::panic;

const BUG_REPORT_URL: &str = "https://github.com/ethereum/fe/issues/new";
static DEFAULT_PANIC_HOOK: Lazy<Box<dyn Fn(&panic::PanicInfo<'_>) + Sync + Send + 'static>> =
    Lazy::new(|| {
        let hook = panic::take_hook();
        panic::set_hook(Box::new(|info| report_ice(info)));
        hook
    });

pub fn install_compiler_panic_hook() {
    Lazy::force(&DEFAULT_PANIC_HOOK);
}

/// Errors can either be an object or static reference.
#[derive(Debug)]
pub enum ErrorKind {
    Str(Cow<'static, str>),
    Analyzer(SemanticError),
    Parser(Vec<Diagnostic>),
}

/// List of errors encountered during compilation.
#[derive(Debug)]
pub struct CompileError {
    pub errors: Vec<ErrorKind>,
}

impl Default for CompileError {
    fn default() -> Self {
        Self::new()
    }
}

impl CompileError {
    pub fn new() -> Self {
        Self { errors: Vec::new() }
    }

    /// Create a single error with a static string.
    pub fn static_str(val: &'static str) -> Self {
        Self {
            errors: vec![ErrorKind::Str(val.into())],
        }
    }

    /// Create a single error with a string object.
    pub fn str(val: &str) -> Self {
        Self {
            errors: vec![ErrorKind::Str(val.to_string().into())],
        }
    }

    pub fn analyzer(err: SemanticError) -> Self {
        Self {
            errors: vec![ErrorKind::Analyzer(err)],
        }
    }
}

impl<'a> From<serde_json::error::Error> for CompileError {
    fn from(_: serde_json::error::Error) -> Self {
        CompileError::static_str("JSON serialization error")
    }
}

impl<'a> From<ethabi::Error> for CompileError {
    fn from(e: ethabi::Error) -> Self {
        CompileError::str(&format!("ethabi error: {}", e))
    }
}

impl<'a> From<SemanticError> for CompileError {
    fn from(e: SemanticError) -> Self {
        CompileError::analyzer(e)
    }
}

fn report_ice(info: &panic::PanicInfo) {
    (*DEFAULT_PANIC_HOOK)(info);

    eprintln!();
    eprintln!("You've hit an internal compiler error. This is a bug in the Fe compiler.");
    eprintln!("Fe is still under heavy development, and isn't yet ready for production use.");
    eprintln!();
    eprintln!("If you would, please report this bug at the following URL:");
    eprintln!("  {}", BUG_REPORT_URL);
}
