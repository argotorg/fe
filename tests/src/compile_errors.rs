//! Tests for contracts that should cause compile errors

use fe_analyzer::errors::{ErrorKind::*, SemanticError};
use fe_common::diagnostics::print_diagnostics;
use fe_common::files::FileStore;
use fe_compiler::errors::ErrorKind;
use rstest::rstest;

#[rstest(
    fixture_file,
    expected_error,
    case("assert_reason_not_string.fe", TypeError),
    case("bad_map.fe", MapTypeError),
    case("bad_map2.fe", MapTypeError),
    case("bad_map3.fe", MapTypeError),
    case("bad_map4.fe", MapTypeError),
    case("bad_map5.fe", TypeError),
    case("break_without_loop_2.fe", BreakWithoutLoop),
    case("break_without_loop.fe", BreakWithoutLoop),
    case("call_event_with_wrong_types.fe", TypeError),
    case("call_undefined_function_on_contract.fe", UndefinedValue),
    case("call_undefined_function_on_external_contract.fe", UndefinedValue),
    case("call_undefined_function_on_memory_struct.fe", UndefinedValue),
    case("call_undefined_function_on_storage_struct.fe", UndefinedValue),
    case("circular_dependency_create.fe", CircularDependency),
    case("circular_dependency_create2.fe", CircularDependency),
    case("continue_without_loop_2.fe", ContinueWithoutLoop),
    case("continue_without_loop.fe", ContinueWithoutLoop),
    case("duplicate_contract_in_module.fe", AlreadyDefined),
    case("duplicate_event_in_contract.fe", AlreadyDefined),
    case("duplicate_field_in_contract.fe", AlreadyDefined),
    case("duplicate_field_in_struct.fe", AlreadyDefined),
    case("duplicate_method_in_contract.fe", AlreadyDefined),
    case("duplicate_struct_in_module.fe", AlreadyDefined),
    case("duplicate_typedef_in_module.fe", AlreadyDefined),
    case("duplicate_var_in_child_scope.fe", AlreadyDefined),
    case("duplicate_var_in_contract_method.fe", AlreadyDefined),
    case("emit_undefined_event.fe", MissingEventDefinition),
    case("external_call_type_error.fe", TypeError),
    case("external_call_wrong_number_of_params.fe", WrongNumberOfParams),
    case("indexed_event.fe", MoreThanThreeIndexedParams),
    case("keccak_called_with_wrong_type.fe", TypeError),
    case("mismatch_return_type.fe", TypeError),
    case("missing_return_in_else.fe", MissingReturn),
    case("missing_return.fe", MissingReturn),
    case("needs_mem_copy.fe", CannotMove),
    case("non_bool_and.fe", TypeError),
    case("non_bool_or.fe", TypeError),
    case("not_in_scope_2.fe", UndefinedValue),
    case("not_in_scope.fe", UndefinedValue),
    case("numeric_capacity_mismatch/i128_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i128_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i16_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i16_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i256_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i256_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i32_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i32_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i64_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i64_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i8_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/i8_pos.fe", NumericCapacityMismatch),
    case(
        "numeric_capacity_mismatch/literal_too_big.fe",
        NumericCapacityMismatch
    ),
    case(
        "numeric_capacity_mismatch/literal_too_small.fe",
        NumericCapacityMismatch
    ),
    case("numeric_capacity_mismatch/u128_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u128_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u16_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u16_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u256_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u256_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u32_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u32_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u64_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u64_pos.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u8_neg.fe", NumericCapacityMismatch),
    case("numeric_capacity_mismatch/u8_pos.fe", NumericCapacityMismatch),
    case("pow_with_signed_exponent.fe", SignedExponentNotAllowed),
    case("pow_with_wrong_capacity.fe", TypeError),
    case("return_addition_with_mixed_types.fe", TypeError),
    case("return_call_to_fn_with_param_type_mismatch.fe", TypeError),
    case("return_call_to_fn_without_return.fe", TypeError),
    case("return_from_init.fe", TypeError),
    case("return_lt_mixed_types.fe", TypeError),
    case("strict_boolean_if_else.fe", TypeError),
    case("string_capacity_mismatch.fe", StringCapacityMismatch),
    case("struct_call_without_kw_args.fe", KeyWordArgsRequired),
    case("type_constructor_from_variable.fe", NumericLiteralExpected),
    case("unary_minus_on_bool.fe", TypeError),
    case("unexpected_return.fe", TypeError),
    case("aug_assign_non_numeric.fe", TypeError),
    case("binary_operations/add_uints.fe", TypeError),
    case("binary_operations/lshift_bool.fe", TypeError),
    case("binary_operations/lshift_with_int.fe", TypeError),
    case("binary_operations/pow_int.fe", TypeError),
    case("bad_tuple_attr1.fe", UndefinedValue),
    case("bad_tuple_attr2.fe", UndefinedValue)
)]
fn test_compile_errors(fixture_file: &str, expected_error: fe_analyzer::errors::ErrorKind) {
    let mut files = FileStore::new();
    let (src, id) = files
        .load_file(&format!("fixtures/compile_errors/{}", fixture_file))
        .expect("unable to read fixture file");

    match fe_compiler::compile(&src, id, true, false) {
        Err(error) => {
            if error.errors.iter().any(|err| match err {
                ErrorKind::Analyzer(SemanticError { kind, .. }) => *kind == expected_error,
                _ => false,
            }) {
                return;
            }

            for err in error.errors {
                match err {
                    ErrorKind::Str(errstr) => eprintln!("Compiler error: {}", errstr),
                    ErrorKind::Analyzer(err) => {
                        eprintln!("Analyzer error: {}", err.format_user(&src))
                    }
                    ErrorKind::Parser(diags) => print_diagnostics(&diags, &files),
                }
            }
            panic!(
                "Expected error {:?} was not found. Unexpected error(s) were.",
                expected_error
            )
        }
        _ => panic!(
            "Compiling succeeded when it was expected to fail with: {:?}",
            expected_error
        ),
    }
}
