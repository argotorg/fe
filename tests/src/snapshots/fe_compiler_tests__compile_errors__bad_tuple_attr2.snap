---
source: tests/src/compile_errors.rs
expression: "error_string(&path, &src)"

---
error: Invalid attempt to access tuple field
  ┌─ fixtures/compile_errors/bad_tuple_attr2.fe:6:27
  │
6 │         self.my_sto_tuple.m
  │                           ^ Invalid attempt
  │
  = Note: Tuple values are accessed via `itemN` properties such as `item0` or `item1`


